compiler/generator_rust.c

```diff
@@ -0,0 +1,1347 @@
+    generate_AE(g, p->AE);
+        w(g, ")");
+    write_block_start(g);
+    write_failure(g);
+    write_block_end(g);
+    g->unreachable = false;
+}
+
+static void generate_call(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    g->V[0] = p->name;
+    write_failure_if(g, "!~W0(env, context)", p);
+}
+
+static void generate_grouping(struct generator * g, struct node * p, int complement) {
+
+    struct grouping * q = p->name->grouping;
+    g->S[0] = p->mode == m_forward ? "" : "_b";
+    g->S[1] = complement ? "out" : "in";
+    g->V[0] = p->name;
+    g->I[0] = q->smallest_ch;
+    g->I[1] = q->largest_ch;
+    write_failure_if(g, "!env.~S1_grouping~S0(~W0, ~I0, ~I1)", p);
+}
+
+static void generate_namedstring(struct generator * g, struct node * p) {
+
+    write_comment(g, p);
+    g->S[0] = p->mode == m_forward ? "" : "_b";
+    g->V[0] = p->name;
+    write_failure_if(g, "!env.eq_s~S0(&~V0)", p);
+}
+
+static void generate_literalstring(struct generator * g, struct node * p) {
+
+    symbol * b = p->literalstring;
+    write_comment(g, p);
+    g->S[0] = p->mode == m_forward ? "" : "_b";
+    g->L[0] = b;
+    write_failure_if(g, "!env.eq_s~S0(&~L0)", p);
+}
+
+static void generate_setup_context(struct generator * g) {
+  struct name * q;
+  w(g, "~Mlet mut context = &mut Context {~+~N");
+  for (q = g->analyser->names; q; q = q->next) {
+    g->V[0] = q;
+    switch (q->type) {
+    case t_string:
+      w(g, "~M~W0: String::new(),~N");
+      break;
+    case t_integer:
+      w(g, "~M~W0: 0,~N");
+      break;
+    case t_boolean:
+      w(g, "~M~W0: false,~N");
+      break;
+    }
+  }
+  w(g, "~-~M};~N");
+}
+
+
+static void generate_define(struct generator * g, struct node * p) {
+
+    struct name * q = p->name;
+
+    struct str * saved_output = g->outbuf;
+
+    g->V[0] = q;
+
+    if (q->type == t_routine) {
+        w(g, "~N~Mfn ~W0(env: &mut SnowballEnv, context: &mut Context) -> bool {~+~N");
+    } else {
+        w(g, "~N~Mpub fn ~W0(env: &mut SnowballEnv) -> bool {~+~N");
+        generate_setup_context(g);
+    }
+    // w(g, "~Mprintln!(\"~W0: \\t\\t\\t{:?}\", env);~N");
+
+    g->outbuf = str_new();
+
+    g->next_label = 0;
+    g->var_number = 0;
+
+    str_clear(g->failure_str);
+    g->failure_label = x_return;
+    g->unreachable = false;
+    generate(g, p->left);
+    if (!g->unreachable) w(g, "~Mreturn true;~N");
+    w(g, "~-~M}~N");
+
+    str_append(saved_output, g->outbuf);
+    str_delete(g->outbuf);
+    g->outbuf = saved_output;
+}
+
+static void generate_substring(struct generator * g, struct node * p) {
+
+    struct among * x = p->among;
+
+    write_comment(g, p);
+
+    g->S[0] = p->mode == m_forward ? "" : "_b";
+    g->I[0] = x->number;
+
+    if (x->command_count == 0 && x->starter == 0) {
+        write_failure_if(g, "env.find_among~S0(~A_~I0, context) == 0", p);
+    } else {
+        writef(g, "~Mlet among_var = env.find_among~S0(~A_~I0, context);~N", p);
```

Comment:
`among_var` does not live long enough in the Rust source generated from
```snowball
externals (stem)
define stem as (
    [gopast substring]
    among('x' (<- 'X'))
)
```
See `generate_define` in generator.c for how to use `amongvar_needed` to fix this.